xarray/backends/zarr.py

```diff
@@ -0,0 +1,554 @@
+    dimensions = zarr_obj.attrs.get(dimension_key)
+    attributes = HiddenKeyDict(zarr_obj.attrs, dimension_key)
+    return dimensions, attributes
+
+
+def _extract_zarr_variable_encoding(variable, raise_on_invalid=False):
+    encoding = variable.encoding.copy()
+
+    valid_encodings = set(['chunks', 'compressor', 'filters',
+                           'cache_metadata'])
+
+    if raise_on_invalid:
+        invalid = [k for k in encoding if k not in valid_encodings]
+        if invalid:
+            raise ValueError('unexpected encoding parameters for zarr '
+                             'backend:  %r' % invalid)
+    else:
+        for k in list(encoding):
+            if k not in valid_encodings:
+                del encoding[k]
+
+    chunks = _determine_zarr_chunks(encoding.get('chunks'), variable.chunks,
+                                    variable.ndim)
+    encoding['chunks'] = chunks
+
+    # TODO: figure out how to serialize compressor and filters options
+    # in zarr these are python objects, not strings
+
+    return encoding
+
+# copied from conventions.encode_cf_variable
+
+def encode_zarr_variable(var, needs_copy=True, name=None):
+    """
+    Converts an Variable into an Variable which follows some
+    of the CF conventions:
+
+        - Nans are masked using _FillValue (or the deprecated missing_value)
+        - Rescaling via: scale_factor and add_offset
+        - datetimes are converted to the CF 'units since time' format
+        - dtype encodings are enforced.
+
+    Parameters
+    ----------
+    var : xarray.Variable
+        A variable holding un-encoded data.
+
+    Returns
+    -------
+    out : xarray.Variable
+        A variable which has been encoded as described above.
+    """
+    var = conventions.maybe_encode_datetime(var, name=name)
+    var = conventions.maybe_encode_timedelta(var, name=name)
+    var, needs_copy = conventions.maybe_encode_offset_and_scale(var,
+                                                        needs_copy, name=name)
+    var, needs_copy = conventions.maybe_encode_fill_value(var, needs_copy,
+                                                          name=name)
+    var = conventions.maybe_encode_nonstring_dtype(var, name=name)
+    var = conventions.maybe_default_fill_value(var)
+    var = conventions.maybe_encode_bools(var)
+    var = conventions.ensure_dtype_not_object(var, name=name)
+    var = conventions.maybe_encode_string_dtype(var, name=name)
+    return var
+
+
+class ZarrStore(AbstractWritableDataStore, DataStorePickleMixin):
+    """Store for reading and writing data via zarr
+    """
+
+    def __init__(self, store=None, mode='a', synchronizer=None, group=None,
+                 writer=None, autoclose=None):
+        self._mode = mode
+        self._synchronizer = synchronizer
+        self._group = group
+
+        # zarr stores don't need to be opened, closed, or synced.
+        # So what do we do with all this logical about openers?
```

Comment:
Can you pickle all zarr objects? If so, we should be able to drop autoclose/opener/DataStorePickleMixin.

Reply:
[Zarr objects (Array, Group) can be pickled](http://zarr.readthedocs.io/en/master/tutorial.html#pickle-support) as long as the underlying store can be pickled (which is true for all store classes provided by zarr).

Reply:
I wonder about `gcsfs.mapping.GCSMap`

Reply:
I believe that @martindurant made sure to test serializability on all of the cloud file system mappings.  Here is a test from that repository:

```python
@my_vcr.use_cassette(match=['all'])
def test_map_pickle(token_restore):
    with gcs_maker() as gcs:
        d = GCSMap(root, gcs)
        d['x'] = b'1'
        assert d['x'] == b'1'

        import pickle
        d2 = pickle.loads(pickle.dumps(d))

        assert d2['x'] == b'1'
```

Reply:
If any of the filesystems or mappers don't pickle successfully, that is a bug and we should fix it. Azure-datalake can be supported too, but that is not yet public.